editor/plugins/camera_2d_editor_plugin.cpp

```diff
@@ -43,36 +43,196 @@ void Camera2DEditor::edit(Camera2D *p_camera) {
+				plugin->update_overlays();
+			} break;
+
+			case Drag::CENTER: {
+				Rect2 target_rect = selected_camera->get_limit_rect();
+				target_rect.position = pos - center_drag_point;
+				selected_camera->set_limit_rect(target_rect);
+				plugin->update_overlays();
+			} break;
+
+			case Drag::NONE: {
+			} break;
+		}
+		return true;
+	}
+
+	return false;
+}
+
+void Camera2DEditor::forward_canvas_draw_over_viewport(Control *p_overlay) {
+	if (!selected_camera || !selected_camera->is_limit_enabled()) {
+		return;
+	}
+	Rect2 limit_rect = selected_camera->get_limit_rect();
+	limit_rect = CanvasItemEditor::get_singleton()->get_canvas_transform().xform(limit_rect);
+	p_overlay->draw_rect(limit_rect, Color(1, 1, 0.25, 0.63), false, 3);
 }
 
 void Camera2DEditor::_menu_option(int p_option) {
 	switch (p_option) {
 		case MENU_SNAP_LIMITS_TO_VIEWPORT: {
 			EditorUndoRedoManager *ur = EditorUndoRedoManager::get_singleton();
-			Rect2 prev_rect = selected_camera->get_limit_rect();
-			ur->create_action(TTR("Snap the Limits to the Viewport"), UndoRedo::MERGE_DISABLE, selected_camera);
-			ur->add_do_method(this, "_snap_limits_to_viewport");
-			ur->add_do_reference(selected_camera);
-			ur->add_undo_method(this, "_undo_snap_limits_to_viewport", prev_rect);
+			ur->create_action(TTR("Snap Camera2D Limits to the Viewport"), UndoRedo::MERGE_DISABLE, selected_camera);
+			ur->add_do_method(this, "_snap_limits_to_viewport", selected_camera);
+			ur->add_undo_method(selected_camera, "_set_limit_rect", selected_camera->get_limit_rect());
+			ur->add_undo_method(this, "_update_overlays_if_needed", selected_camera);
 			ur->commit_action();
 		} break;
 	}
 }
 
-void Camera2DEditor::_snap_limits_to_viewport() {
-	selected_camera->set_limit(SIDE_LEFT, 0);
-	selected_camera->set_limit(SIDE_TOP, 0);
-	selected_camera->set_limit(SIDE_RIGHT, GLOBAL_GET("display/window/size/viewport_width"));
-	selected_camera->set_limit(SIDE_BOTTOM, GLOBAL_GET("display/window/size/viewport_height"));
+void Camera2DEditor::_snap_limits_to_viewport(Camera2D *p_camera) {
+	p_camera->set_limit(SIDE_LEFT, 0);
+	p_camera->set_limit(SIDE_TOP, 0);
+	p_camera->set_limit(SIDE_RIGHT, GLOBAL_GET("display/window/size/viewport_width"));
+	p_camera->set_limit(SIDE_BOTTOM, GLOBAL_GET("display/window/size/viewport_height"));
+	_update_overlays_if_needed(p_camera);
+}
+
+void Camera2DEditor::_update_overlays_if_needed(Camera2D *p_camera) {
+	if (p_camera == selected_camera) {
+		plugin->update_overlays();
+	}
 }
 
-void Camera2DEditor::_undo_snap_limits_to_viewport(const Rect2 &p_prev_rect) {
-	Point2 end = p_prev_rect.get_end();
-	selected_camera->set_limit(SIDE_LEFT, p_prev_rect.position.x);
-	selected_camera->set_limit(SIDE_TOP, p_prev_rect.position.y);
-	selected_camera->set_limit(SIDE_RIGHT, end.x);
-	selected_camera->set_limit(SIDE_BOTTOM, end.y);
+void Camera2DEditor::_update_hover(const Vector2 &p_mouse_pos) {
```

Comment:
When the tool changes in `CanvasItemEditor::_button_tool_select`, `_update_hover` should be called somehow. Because the tool can change while hovering over a limit and it won't update until the mouse is moved.
This is a pretty small issue though, its not blocking.